### ISSUE TITLE
vertex lighting artifacts

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#291]
noticed in #289 and #290 

introduced in 4ea413240a2593f44f71c2d4b45f48496b4e77db by @gimhael

before 4ea4132:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_001635_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_001635_000.jpg)

after 4ea4132:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_001519_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_001519_000.jpg)

**Comment by illwieckz** [comment]
Since the commit introducing the bug is unifying the code for texture with or without normal map and that map does have normal map, the issue probably only occurs with simple materials without normal maps and maybe the code is not introduced but uncovered.

**Comment by illwieckz** [comment]
@gimhael, is it something you can investigate?

**Comment by illwieckz** [comment]
Test map: https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/map-metro_0~b1-2.dpk

**Comment by DolceTriade** [comment]
Is vertex lighting something we need to support anymore?

**Comment by illwieckz** [comment]
For sure, world vertex lighting is pretty useless (though I'm impressed about how not-that-bad it looks). For arena games, hardcore gamers may still want it to “simplify” the view (no complex shadows means less information on screen to process).

Vertex lighting is still used for entity, though the whole code is currently copy-pasted in a separate glsl file.

I have currently no way to know if this bug may never affects entity lighting. Without knowing what is wrong with vertex world lighting we have to consider this bug may affect models as well, as soon they reproduce the same conditions. Looking for the cause of this bug it the best way to determine if models can be affected.

I have a work-in-progress branch to merge vertex entity / vertex world / lightmap world in one code (with related code switches around the only differences instead of duplicating code everywhere), while I'm not against dropping it, vertex world lighting feature would be almost free to keep alive.

My main concern is to be sure models will never be affected by this.

**Comment by illwieckz** [comment]
Entities are affected too, I used an old tremulous weapon (chaingun) and while I move using `noclip` and the light of the model is going wrongly dark or light at the same time the ceiling is wrongly going dark or light:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022640_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022640_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022650_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022650_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022701_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022701_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022707_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_022707_000.jpg)

**Comment by illwieckz** [comment]
I now remember having noticed this long time ago, by floating around in this map…

**Comment by illwieckz** [comment]
This is also affecting multitextured materials, because I reproduce the bug with Unvanquished models:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_023403_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_023403_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_023407_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_023407_000.jpg)

**Comment by illwieckz** [comment]
Note that, since I do not reproduce the bug with our own maps and this one was built more than 10 years ago,  I **do not exclude** the hypothesis of the map having wrong information in BSP because of an old, now-fixed `q3map2` bug. If this is true, we would still have make efforts to be sure to render this map build properly because :

1. many other maps from that time may reproduce the bug,
2. we managed to render them properly up to 0.50, so _we can do it_.

**Comment by illwieckz** [comment]
The bug can be seen on wall-walking aliens, the lighting is going strong and dim wrongly on the dretch like on the ceiling:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_024800_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_024800_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_024811_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_024811_000.jpg)

It's visible on the alien model while the world is rendered with lightmaps:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025710_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025710_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025718_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025718_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025813_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025813_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025822_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-02-26_025822_000.jpg)

**Comment by illwieckz** [comment]
If I add:

```glsl
	lightDir = normal;
```

before this line, the weird artifacts are gone:

```glsl
	computeLight(lightDir, normal, viewDir, lightColor, diffuse, material, color);
```

That was basically what was done for materials without normal maps before 4ea4132 (other part of that old code seems to have no effect).

Currently, `lightDir` is computed from `lightGrid1.a` and `lightGrid1.b` by `ReadLightGrid()`.
Maybe this data is wrong or undefined in this map. Can it be related to deluxe mapping? This map does not use deluxe mapping.

**Comment by illwieckz** [comment]
`ReadLightGrid()` does this:

```glsl
	vec4 texel1 = texture3D(u_LightGrid1, pos);
	vec4 texel2 = texture3D(u_LightGrid2, pos);

	lightDir.x = (255.0 * texel1.w - 128.0) / 127.0;
	lightDir.y = (255.0 * texel2.w - 128.0) / 127.0;
	lightDir.z = 1.0 - abs(lightDir.x) - abs(lightDir.y);
```

`u_LightGrid1` and `u_LightGrid2` are lumps of things defined that way (as seen in `tr_local.h`):

```
	// The light direction vector is stored as the x/y coordinates
	// of the vector projected on an unit octahedron. To disambiguate
	// the upper and lower half of the octahedron, the four lower
	// triangles of the octahedron are flipped into the outer corners
	// of the unit square.
	struct bspGridPoint1_t
	{
		byte  ambient[3];
		byte  lightVecX;
	};
	struct bspGridPoint2_t
	{
		byte  directed[3];
		byte  lightVecY;
	};
```

While [Q3 BSP format](http://www.mralligator.com/q3/#Lightvols) define the same things this way:

```
ubyte[3] ambient     | Ambient color component.      RGB.
ubyte[3] directional | Directional color component.  RGB.
ubyte[2] dir         | Direction to light.           0=phi, 1=theta.
```

I don't know what this _phi_ and _theta_ are about. Is that data _the same_ at first?

**Comment by illwieckz** [comment]
So, phi and theta look to be [defined this way](https://mathinsight.org/spherical_coordinates):

[![phi, theta, x, y](https://mathinsight.org/media/image/image/spherical_coordinates.png)](https://mathinsight.org/spherical_coordinates)

[![phi, theta, x, y](https://mathinsight.org/media/image/image/spherical_coordinates_cartesian.png)](https://mathinsight.org/spherical_coordinates)

And there is this code in `tr_bsp.cpp`:

```c++
		// standard spherical coordinates to cartesian coordinates conversion

		// decode X as cos( lat ) * sin( long )
		// decode Y as sin( lat ) * sin( long )
		// decode Z as cos( long )

		// RB: having a look in NormalToLatLong used by q3map2 shows the order of latLong

		// Lat = 0 at (1,0,0) to 360 (-1,0,0), encoded in 8-bit sine table format
		// Lng = 0 at (0,0,1) to 180 (0,0,-1), encoded in 8-bit sine table format

		lat = DEG2RAD( in->latLong[ 1 ] * ( 360.0f / 255.0f ) );
		lng = DEG2RAD( in->latLong[ 0 ] * ( 360.0f / 255.0f ) );

		direction[ 0 ] = cos( lat ) * sin( lng );
		direction[ 1 ] = sin( lat ) * sin( lng );
		direction[ 2 ] = cos( lng );

		// Pack data into an bspGridPoint
		gridPoint1->ambient[ 0 ] = floatToUnorm8( ambientColor[ 0 ] );
		gridPoint1->ambient[ 1 ] = floatToUnorm8( ambientColor[ 1 ] );
		gridPoint1->ambient[ 2 ] = floatToUnorm8( ambientColor[ 2 ] );
		gridPoint2->directed[ 0 ] = floatToUnorm8( directedColor[ 0 ] );
		gridPoint2->directed[ 1 ] = floatToUnorm8( directedColor[ 1 ] );
		gridPoint2->directed[ 2 ] = floatToUnorm8( directedColor[ 2 ] );

		// Light direction vectors have to be stored in two bytes:
		// First the vector is projected onto a unit octahedron, that means |x| + |y| + |z| = 1,
		// then it is projected onto the x/y plane. The magnitude of z can be reconstructed by
		// the above identity, but not the sign.
		// Fortunately the identity implies |x| + |y| <= 1, so all vectors fall within a diamond
		// shape within the unit square that covers exactly half of the area:
		//
		//           +-----+-----+
		//           |    /|\    |
		//           |   /#|#\   |
		//           |  /##|##\  |
		//           | /###|###\ |
		//           |/####|####\|
		//           +-----+-----+
		//           |\####|####/|
		//           | \###|###/ |
		//           |  \##|##/  |
		//           |   \#|#/   |
		//           |    \|/    |
		//           +-----+-----+
		//
		// If z >= 0, we keep just the x,y coordinates in the diamond, otherwise the point
		// is flipped across the nearest diamond edge into one of the outer triangles.

		// The interpolation in this format behaves quite good except when interpolating
		// two points that are in different outer triangles.

		scale = fabsf( direction[ 0 ] ) + fabsf( direction[ 1 ] ) + fabsf( direction[ 2 ] );
		if( scale > 0.0f ) {
			VectorScale( direction, 1.0f / scale, direction );
			if( direction[ 2 ] < 0.0f ) {
				float X = direction[ 0 ];
				float Y = direction[ 1 ];
				direction[ 0 ] = copysignf( 1.0f - fabs( Y ), X );
				direction[ 1 ] = copysignf( 1.0f - fabs( X ), Y );
			}
		}
		gridPoint1->lightVecX = 128 + floatToSnorm8( direction[ 0 ] );
		gridPoint2->lightVecY = 128 + floatToSnorm8( direction[ 1 ] );
	}

	// fill in gridpoints with zero light (samples in walls) to avoid
	// darkening of objects near walls
	gridPoint1 = w->lightGridData1;
	gridPoint2 = w->lightGridData2;

	for( k = 0; k < w->lightGridBounds[ 2 ]; k++ ) {
		from[ 2 ] = k - 1;
		to[ 2 ] = k + 1;

		for( j = 0; j < w->lightGridBounds[ 1 ]; j++ ) {
			from[ 1 ] = j - 1;
			to[ 1 ] = j + 1;

			for( i = 0; i < w->lightGridBounds[ 0 ];
			     i++, gridPoint1++, gridPoint2++ ) {
				from[ 0 ] = i - 1;
				to[ 0 ] = i + 1;

				if( gridPoint1->ambient[ 0 ] ||
				    gridPoint1->ambient[ 1 ] ||
				    gridPoint1->ambient[ 2 ] )
					continue;

				scale = R_InterpolateLightGrid( w, from, to, factors,
								ambientColor, directedColor,
								direction );
				if( scale > 0.0f ) {
					scale = 1.0f / scale;

					VectorScale( ambientColor, scale, ambientColor );
					VectorScale( directedColor, scale, directedColor );
					VectorScale( direction, scale, direction );

					gridPoint1->ambient[ 0 ] = floatToUnorm8( ambientColor[ 0 ] );
					gridPoint1->ambient[ 1 ] = floatToUnorm8( ambientColor[ 1 ] );
					gridPoint1->ambient[ 2 ] = floatToUnorm8( ambientColor[ 2 ] );
					gridPoint1->lightVecX = 128 + floatToSnorm8( direction[ 0 ] );
					gridPoint2->directed[ 0 ] = floatToUnorm8( directedColor[ 0 ] );
					gridPoint2->directed[ 1 ] = floatToUnorm8( directedColor[ 1 ] );
					gridPoint2->directed[ 2 ] = floatToUnorm8( directedColor[ 2 ] );
					gridPoint2->lightVecY = 128 + floatToSnorm8( direction[ 1 ] );
				}
			}
		}
	}
```

**Comment by gimhael** [comment]
For vertex lighting the BSP format stores the incoming light premultiplied with the cosine of the angle between the vertex normal and the (average) light direction as (r,g,b) triple. This term sufficient for diffuse lighting, but if you want normalmapping you need directional lighting.

For dynamic objects the BSP file contains the light grid, which stores the ambient and directional components of the incoming light as (r,g,b) triples plus the incoming light direction as two angles as described above. Now the idea was to use the light grid instead of vertex light for normal mapping.

This approach has some problems: The light grid sa
Thus the mples only a finite number of points, in between the data is interpolated. If two adjacent sample points are on the opposite sides of a wall the interpolation can get wrong results. Or a sample point could lie inside a wall where no light information is available.

These kind of problems cannot be solved by the renderer, they have to be dealt with by recompiling the map with a sufficient light grid size.

Then there are other problems: reading light grid data from a 3d texture is easy in GLSL, but it always performs trilinear interpolation of the data. While this is wanted for the light colours, interpolation of angles can lead to unrealistic results. Also converting angles to vectors requires very expensive sin and cos instructions, which should be avoided in a fragment shader.
It is much better to convert the angles into normal vectors and then interpolate those vectors, however then we end up with 9 components (2x(r,g,b) and 1x(x,y,z) for the direction), so they don't fit into two textures.

The octahedron compression described in the source code is used to store the direction in two components as with the polar format while still being reasonable fast to decode and good behaving when interpolated.
However the interpolation can still fail if the light direction is mostly from below (if the light vectors of two adjacent grid points fall into different corners of the diagram above, the interpolation path will go through the shaded inner part, which means the interpolation of two downward vectors can yield an upward vector).

I think that can only be fixed if we store the full direction vector in the light grid, so we would have to either store three light grid textures or to use some form of compression for the two (r,g,b) components.

**Comment by illwieckz** [comment]
Hi, thank you for your answer.

Before answering in topic, do you mean?

```diff
- This approach has some problems: The light grid sa
- Thus the mples only a finite number of points, in between the data is interpolated.
+ This approach has some problems: The light grid samples only a finite number
+ of points, in between the data is interpolated.
```

Then, in topic:

> Now the idea was to use the light grid instead of vertex light for normal mapping.

Are you meaning our engine does not do vertex lighting at all but does lightgrid lighting despite this feature is called “vertex lighting”? That may explain why I'm surprised how that good looks what we call “vertex lighting” in dæmon.

> These kind of problems cannot be solved by the renderer, they have to be dealt with by recompiling the map with a sufficient light grid size.

Is there any way to somewhat guess the light grid size is too small (if possible at map load)?

> Also converting angles to vectors requires very expensive sin and cos instructions, which should be avoided in a fragment shader.
It is much better to convert the angles into normal vectors and then interpolate those vectors, however then we end up with 9 components (2x(r,g,b) and 1x(x,y,z) for the direction), so they don't fit into two textures.

How much this would cost in rendering performance to bind one texture more?

> Or a sample point could lie inside a wall where no light information is available. […] Then there are other problems: […] I think that can only be fixed if we store the full direction vector in the light grid , so we would have to either store three light grid textures.

Are you telling that if a sample point lies inside a wall, the 3 texture light grid trick would be useless anyway?

I'm secretly hoping to get that lightgrid trick working in all case so we can enable normal mapping on old map without sources, just by shipping normalmaps…

**Comment by gimhael** [comment]
Yes, the engine always uses the light grid if no light maps are available. The vertex lights stored in
the BSP are not used at all.

An additional texture fetch is probably no issue on DX10/OpenGL3 or higher hardware, but in DX9/OpenGL2 the number of texture mapping units is limited.

If a light grid sample falls into a wall (or otherwise outside the map), there BSP contains a zero light value at that position. When the texture interpolation mixes those black samples with samples from inside the map, the resulting colour will be darker than natural.

The code in R_LoadLightgrid tries to fill those black samples with an interpolated colour, at least in the cases where black samples are adjacent to non-black samples, so in practice this should not be an issue.

**Comment by gimhael** [comment]
I have implemented a different light grid texture format at https://github.com/gimhael/Daemon/tree/lightgrid, could you check out if this improves the lighting on the map.

**Comment by illwieckz** [comment]
Thank you ve much! this is very precious! I'll try it tomorrow when I'm at
my computer again. =)

Note that the test map is there:
https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/map-metro_0~b1-2.dpk

I've seen you also rewrote the lightgrid code in liquid shader, which was
very different and looked to be wrong by the way. So you implicitly answer
to another question I haven't asked yet: can we use the same lightgrid code
in every shader? Yes.

**Comment by illwieckz** [comment]
@gimhael , Man! You're a true god! my savior! (almost :grin:)

This is lightigrid lighting

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_214744_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_214744_000.jpg)

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_214955_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_214955_000.jpg)

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215109_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215109_000.jpg)

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215112_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215112_000.jpg)

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215149_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215149_000.jpg)

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_220117_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_220117_000.jpg)

Note, due to bug #296, those two screenshots display both lightgrid lighting and lightmap lighting on the same screen. Lightmap on the descending stairs on the right and lightigrid on the left platform between staircases:

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215039_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215039_000.jpg)

Lightmap on the stairs and on the wall with “Titanic 2” movie poster parody on the left, lightgrid everywhere else on the left corridor and background station platform:

[![lightgrid lighting](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215054_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_215054_000.jpg)

Note that there was a minor copy paste issue in liquid shader, that was easy to fix.

**Comment by illwieckz** [comment]
With that feature, it would possible to disable lightmap on moving models, to avoid the bug of shadows moving with the models:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_221653_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_221653_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222322_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222322_000.jpg)

For reference, previous vega screenshot done with lightmap instead:

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222245_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222245_000.jpg)

**Comment by illwieckz** [comment]
With lightgrid then with lightmap (noFast build):

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222114_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222114_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222147_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222147_000.jpg)

With lightgrid then with lightmap (noFast build):

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222426_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222426_000.jpg)

[![vertex lighting artifacts](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222453_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/vertex-lighting-artifacts/unvanquished_2020-03-01_222453_000.jpg)

**I would like to see how looks a baked lightgrid lighting with realtime shadows casted from pointlights / emitting shaders.**

**Comment by illwieckz** [comment]
> **@gimhael** An additional texture fetch is probably no issue on DirectX 10/OpenGL 3.0 or higher hardware, but in DX9/OpenGL2 the number of texture mapping units is limited.

How much of those obsolete drivers may be in the wild? DirectX 10 is as old as Vista (2007), OpenGL 3.0 dates from 2008.

On [AMD side](https://en.wikipedia.org/wiki/List_of_AMD_graphics_processing_units), having an OpenGL 2 driver means having a pre-Radeon HD GPU (before R600), meaning a GPU from before 2005. According to Wikipedia, even my old HD 2600 PRO AGP (yes, AGP) that was on market in 2007 is supposed to support OpenGL 3.3. Radeon HD 2000 that was out in 2007 is told to support DirectX 10 and OpenGL 3.3.

On [NVidia side](https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units), it's harder to tell, but it looks like Geforce 8 (starting to ship in 2006) supports DirectX 10 and OpenGL 3.3. Although, some later models has not like GeForce 9600 GSO and GeForce 9400 mGPU (from 2008) or GeForce GT 230 (from 2009) that seems to only supports OpenGL 2.1 but DirectX 10.

On [Intel side](https://en.wikipedia.org/wiki/List_of_Intel_graphics_processing_units), it looks like supports for DirectX 10 with some Gen4 in 2007 (with OpenGL 2.1 unfortunately), and to get OpenGL 3.1 it was required to wait for 2011 with Gen6 (also supporting DirectX 10.1).

This is for Windows and Linux. For macOS, [This official document by Apple from 2017](https://developer.apple.com/opengl/OpenGL-Capabilities-Tables.pdf) says macOS X 10.7.5 (Lion) supports OpenGL 3.1, which was out in 2011. This [other official Apple page updated in 2019](https://support.apple.com/en-us/HT202823) do not list hardware with less than OpenGL 3.3, even for hardware from year 2007.

So, not having DirectX 10 and OpenGL 3.0 support means decade old hardware. Note that I myself own some of them. For example I have a Thinkpad X61 Tablet that has a GMA X3100 which Wikipedia claims to not support OpenGL 3.3. I still own this, it works, but I don't use it. It would be probably slow to play Unvanquished by the way, even on the 1024×768 screen. This laptop started to sell in 2007. Wikipedia says this GPU provides OpenGL 2.1 on Linux and macOS (if exists in a Mac) and OpenGL 1.5 on Windows so this laptop won't run Unvanquished on Windows (DirectX 10 is supported by this GPU but Dæmon don't support this tech).

If we want to drive tests on old hardware, I have a bunch of old GPUs, like a pre-Radeon HD that is currently orphaned but thanks to long-living PCIe standard I may still plug it in a modern host.

If some feature does not work with OpenGL 2 and we want to keep support for it in engine, I'm in favor of disabling non-required features instead of restraining us to implement required features. For example if we bind too much textures for lightgrid on OpenGL 2, we may drop deluxe maps, normal maps and specular maps on such systems.

So, I would like to know your thoughts about OpenGL 2.

I guess the dark areas with dragons may be found around some decade-old Intel chips running macOS or Windows. Some may still be alive and running around there, even if it's probably a bit slow to run Unvanquished.

**Comment by illwieckz** [comment]
So, to drive OpenGL 2 tests, I have:

- an Intel GMA X3100 (2007, GL 2.1) in a working laptop with up-to-date Linux system
- an ATI Radeon X550 PCIe (2005, GL 2.0) without host
- an ATI Radeon X1050 PCIe (2006, GL 2.0) without host
- an nVidia Geforce 6600 PCIe (2004, GL 2.1) without host
- an nVidia Geforce 7500 LE PCIe (2006, GL 2.1) without host

and maybe some others.

**Comment by illwieckz** [comment]
Fixed by #298.

**Comment by illwieckz** [comment]
In light grid scenario, `lightDir` is computed this way:

```glsl
lightDir = normalize(texel.xyz - (128.0 / 255.0));
```

In deluxe map scenario, `lightDir` is computed this way:

```glsl
lightDir = normalize(2.0 * deluxe.xyz - 1.0);
```

@gimhael, do you think it would be possible to transform the lightgrid direction texture in a way we can use the deluxe code to compute it, without performance loss neither problematic precision loss? This would help to make glsl shaders simpler (especially if one day we merge some of them).
